common/client/node_fsm.go

```diff
@@ -115,15 +119,59 @@ func init() {
 
 // State allows reading the current state of the node.
 func (n *node[CHAIN_ID, HEAD, RPC]) State() nodeState {
+	n.stateMu.RLock()
+	defer n.stateMu.RUnlock()
+	return n.recalculateState()
+}
+
+func (n *node[CHAIN_ID, HEAD, RPC]) getCachedState() nodeState {
 	n.stateMu.RLock()
 	defer n.stateMu.RUnlock()
 	return n.state
 }
 
-func (n *node[CHAIN_ID, HEAD, RPC]) StateAndLatest() (nodeState, int64, *big.Int) {
+func (n *node[CHAIN_ID, HEAD, RPC]) recalculateState() nodeState {
+	if n.state != nodeStateAlive {
+		return n.state
+	}
+
+	// double check that node is not lagging on finalized block
+	if n.nodePoolCfg.EnforceRepeatableRead() && n.isFinalizedBlockOutOfSync() {
+		return nodeStateFinalizedBlockOutOfSync
+	}
+
+	return nodeStateAlive
+}
+
+func (n *node[CHAIN_ID, HEAD, RPC]) isFinalizedBlockOutOfSync() bool {
+	if n.poolInfoProvider == nil {
+		return false
+	}
+
+	highestObservedByCaller := n.poolInfoProvider.HighestUserObservations()
+	latest, _ := n.rpc.GetInterceptedChainInfo()
```

Comment:
Doesn't `poolInfoProvider.HighestUserObservations` already retrieve highest observations among nodes, including this node? Why do we refetch?